### ISSUE TITLE
Question on msl loss in "train_msl.py"

### ISSUE THREAD

**Issue by hsyoo32** [WANGBohaO-jpg/MSL#2]
Hi! I recently read your paper and found it very interesting.
I noticed that in Line 112 of "train_msl.py", pos_logits is not passed through logexp, while in "train_lml".py, it uses logexp of pos_logits for numerator. Could you clarify whether this is an inconsistency, or am I misunderstanding something?

**Comment by WANGBohaO-jpg** [comment]
Thank you very much for your interest in our paper and for your careful examination of both the manuscript and the code!

The MSL loss (Equation 5) is defined in the paper as:

```math
\mathcal{L}_{MSL}(x, y^p; \theta) = \sum_{t=1}^{|y^p|} - \log \frac{\exp(f_\theta(y_t^p))}{\sum_{z \in \mathcal{Z}_{valid}} \exp(f_\theta(z))}
```

This formula can be equivalently expanded as:

```math
\begin{align*}
\mathcal{L}_{MSL}(x, y^p; \theta) 
&= \sum_{t=1}^{|y^p|} \left[ -f_\theta(y_t^p) + \log \sum_{z \in \mathcal{Z}_{valid}} \exp(f_\theta(z)) \right]
\end{align*}
```

Therefore, regarding the positive logits part in "train_msl.py":

```python
pos_logits = shift_logits.gather(1, shift_labels.unsqueeze(1)).squeeze(1)
pos_loss = -pos_logits / self.tau
```

is mathematically equivalent to the implementation in "train_lml.py"

```python
pos_logits = torch.exp(shift_logits.gather(1, shift_labels.unsqueeze(1)).squeeze(1) / self.tau)
pos_loss = -torch.log(pos_logits)
```
In other words, both implementations yield the same result for the positive part of the loss, just expressed in different forms.

**Comment by hsyoo32** [comment]
Thank you for the clarification! 

I have one more question about the theorem. In Equation 15, should the inequality in the denominator be reversed, or should v and p be swapped? Could you please take a look?

![Image](https://github.com/user-attachments/assets/30a142e0-d36e-4c1d-bfcb-f725867a2c24)

**Comment by WANGBohaO-jpg** [comment]
Thank you very much for pointing out the issue in Equation 15. You are absolutely correct—this was a typographical error. The inequality in the denominator should indeed be reversed. I will correct this mistake in the next revision of the manuscript on arXiv. Thank you again for helping to improve the quality of the paper.

**Comment by hsyoo32** [comment]
Thank you very much for your prompt clarification!
